### PR TITLE
Can destructure nested tuples within tuples

### PR DESCRIPTION
**Pull request by matthewmcgarvey**
## Summary

This **ONLY** provides the ability to destructure a tuple inside a tuple.

```mint
component Main {
  fun render : String {
    case ({"A", {"B"}}) {
      {a, {b}} =>
        b
    }
  }
}
```

## Next Steps

The next changes/prs will be to expand the tuple destructuring to support nested enums and arrays. Then I will add the functionality to allow nested destructurings within those as well.

### CODE REVIEW

**Comment by matthewmcgarvey** [comment]
Variable handling was by far the hardest part. It's very ugly but I really wanted to get it working on thought someone else might have a good idea on a better solution.

File: src/compilers/variable.cr [hunk 1]
```diff
@@ -96,12 +96,17 @@ module Mint
...
           case item = entity[0]
           when Ast::WhereStatement, Ast::Statement
             case target = item.target
             when Ast::TupleDestructuring
-              js.variable_of(target.parameters[entity[1]])
+              case val = entity[1]
+              when Int32
+                js.variable_of(target.parameters[val])
+              when Array(Int32)
+                js.variable_of(val.reduce(target) { |curr_type, curr_val| curr_type.as(Ast::TupleDestructuring).parameters[curr_val] })
```
Comment (Sija) [inline src/compilers/variable.cr:108]:
```suggestion
                js.variable_of(val.reduce(target) do |curr_type, curr_val|
                  curr_type.as(Ast::TupleDestructuring).parameters[curr_val]
                end)
```

File: src/errors/top_level.cr [hunk 2]
```diff
@@ -3,7 +3,7 @@ def raise(error : Mint::Error.class)
   raise error, {} of String => Mint::Error::Value
 end
 
-def raise(error : Mint::Error.class, raw : Hash(String, T)) forall T
+def raise(error : Mint::Error.class, raw : Hash(String, Mint::Error::Value?))
```
Comment (Sija) [inline src/errors/top_level.cr:6]: Why is this needed?

**Review by Sija — COMMENTED** [review]
I have just a couple of comments, but besides that looks good!
Is this PR finished btw?

File: src/errors/top_level.cr (hunk 2 above)
Comment (matthewmcgarvey) [reply src/errors/top_level.cr]: It was just a compilation error change I was making along the way. Forgot to revert it.

File: src/compilers/variable.cr [hunk 3]
```diff
@@ -96,12 +96,19 @@ module Mint
           else
             "SHOULD NEVER HAPPEN"
           end
-        when Tuple(Ast::Node, Int32)
+        when Tuple(Ast::Node, Int32), Tuple(Ast::Node, Array(Int32) | Int32)
```
Comment (gdotdesign) [inline src/compilers/variable.cr:99]: `Tuple(Ast::Node, Array(Int32) | Int32)` is enough here since it's matching both cases.

```suggestion
        when Tuple(Ast::Node, Array(Int32) | Int32)
```

File: src/compilers/variable.cr [hunk 4]
```diff
@@ -96,12 +96,19 @@ module Mint
...
             when Ast::TupleDestructuring
-              js.variable_of(target.parameters[entity[1]])
+              case val = entity[1]
+              when Int32
+                js.variable_of(target.parameters[val])
+              when Array(Int32)
+                js.variable_of(val.reduce(target) do |curr_type, curr_val|
+                  curr_type.as(Ast::TupleDestructuring).parameters[curr_val]
+                end)
+              end.not_nil!
```
Comment (gdotdesign) [inline src/compilers/variable.cr:111]: @Sija which is more idiomatic? 

```
end.not_nil!
``` 

or

```
else
  raise "Cannot Happen!"
end
```

File: src/errors/error.cr [hunk 5]
```diff
@@ -4,7 +4,7 @@ module Mint
   class Error < Exception
     alias Value = String | Ast::Node | TypeChecker::Checkable |
```
Comment (gdotdesign) [inline src/errors/error.cr:5]: Like before `Tuple(Ast::Node, Array(Int32) | Int32)` is enough here since it's matching both cases.

File: src/message.cr [hunk 6]
```diff
@@ -176,7 +176,8 @@ module Mint
                                            TypeChecker::Checkable |
                                            Array(TypeChecker::Checkable) |
                                            Array(String) |
-                                           Tuple(Ast::Node, Int32))
+                                           Tuple(Ast::Node, Int32) |
```
Comment (gdotdesign) [inline src/message.cr:179]: Ditto  `Tuple(Ast::Node, Array(Int32) | Int32)`

File: src/type_checkers/scope.cr [hunk 7]
```diff
@@ -12,8 +12,8 @@ module Mint
                    Ast::Store |
                    Ast::Style
 
-      alias Level = Tuple(Ast::Node | Checkable | Tuple(Ast::Node, Int32), Node)
-      alias Lookup = Tuple(Ast::Node | Checkable | Tuple(Ast::Node, Int32), Node, Array(Node))
+      alias Level = Tuple(Ast::Node | Checkable | Tuple(Ast::Node, Int32) | Tuple(Ast::Node, Array(Int32)), Node)
+      alias Lookup = Tuple(Ast::Node | Checkable | Tuple(Ast::Node, Int32) | Tuple(Ast::Node, Array(Int32)), Node, Array(Node))
```
Comment (gdotdesign) [inline src/type_checkers/scope.cr:16]: Ditto  `Tuple(Ast::Node, Array(Int32) | Int32)`.

File: src/type_checkers/statement.cr [hunk 8]
```diff
@@ -6,24 +6,31 @@ module Mint
...
+      } if target.parameters.size > condition.parameters.size
+
+      target.parameters.each_with_index do |param, index|
+        subcondition = condition.parameters[index]
+        _check_statement_target(param, node, subcondition)
       end
+    end
 
-      types[node] = type
+    private def _check_statement_target(target, node, condition)
```
Comment (gdotdesign) [inline src/type_checkers/statement.cr:33]: Is this needed for a `nil` branch?

File: src/type_checkers/variable.cr [hunk 9]
```diff
@@ -47,7 +47,7 @@ module Mint
               type
             end
           end
-        when Tuple(Ast::Node, Int32)
+        when Tuple(Ast::Node, Int32), Tuple(Ast::Node, Array(Int32))
```
Comment (gdotdesign) [inline src/type_checkers/variable.cr:50]: Ditto `Tuple(Ast::Node, Array(Int32) | Int32)`

**Review by gdotdesign — COMMENTED** [review]
There are some minor issues but other than that it's LGTM :+1:

File: src/compilers/variable.cr (hunk 4 above)
Comment (Sija) [reply src/compilers/variable.cr:111]: In cases like these - in which I'm pretty sure the value is there - I'd use `not_nil!` since the stacktrace should be enough (except for cases in which we build mint with `--no-debug` flag - but then we're covered as long as the bug is reproducible).

Comment (gdotdesign) [reply src/compilers/variable.cr:111]: Got it thanks :+1:

File: src/parsers/tuple_destructuring.cr [hunk 10]
```diff
@@ -31,5 +31,9 @@ module Mint
           input: data)
       end
     end
+
+    private def tuple_destructuring_value : Ast::Node?
```
Comment (Sija) [inline src/parsers/tuple_destructuring.cr:35]: I'd call it in the same vein as the other methods, akin to `tuple_destructuring_or_variable`, along with proper `Ast` classes - no casts plz.

File: src/type_checkers/statement.cr (hunk 8 above)
Comment (matthewmcgarvey) [reply src/type_checkers/statement.cr:33]: No, there are multiple types that the where statement target could have but the only one of them we validate is tuple destructuring and that check is also recursive. Rather than do type checking if statements, I use an overload to catch and ignore anything that is not a tuple destructuring.

File: src/type_checkers/variable.cr [hunk 11]
```diff
@@ -57,7 +57,12 @@ module Mint
           when Ast::Statement, Ast::WhereStatement
             case item.target
             when Ast::TupleDestructuring
-              type.parameters[value[1]]
+              case val = value[1]
+              when Int32
+                type.parameters[val]
+              when Array(Int32)
+                val.reduce(type) { |curr_type, curr_val| curr_type.parameters[curr_val] }
+              end.not_nil!
```
Comment (Sija) [inline src/type_checkers/variable.cr:65]:
```suggestion
              in Int32
                type.parameters[val]
              in Array(Int32)
                val.reduce(type) { |curr_type, curr_val| curr_type.parameters[curr_val] }
              end
```

Comment (matthewmcgarvey) [reply src/type_checkers/variable.cr]: Not sure how I've gone this long without knowing about this!

Comment (gdotdesign) [reply src/type_checkers/variable.cr]: Me neither :joy: This needs be documented here https://tw.crystal-lang.org/docs/syntax_and_semantics/case.html :D

File: src/compilers/variable.cr [hunk 12]
```diff
@@ -96,12 +96,19 @@ module Mint
...
```
Comment (gdotdesign) [inline src/compilers/variable.cr:111]: We can use `in` here as well I guess.
```suggestion
              case val = entity[1]
              in Int32
                js.variable_of(target.parameters[val])
              in Array(Int32)
                js.variable_of(val.reduce(target) do |curr_type, curr_val|
                  curr_type.as(Ast::TupleDestructuring).parameters[curr_val]
                end)
              end
```

File: src/type_checkers/statement.cr (hunk 8 above)
Comment (Sija) [inline src/type_checkers/statement.cr:33]:
```suggestion
    private def _check_statement_target(target : Nil, node, condition)
```

**Review by Sija — APPROVED** [review]
LGTM 🚀

**Comment by Sija** [comment]
@matthewmcgarvey Does that supersede #283?

**Comment by matthewmcgarvey** [comment]
@Sija I am not going to be touching that code but I am referencing it. It was really helpful to start with and then I found it was missing a bunch of stuff

**Comment by Sija** [comment]
@matthewmcgarvey What I meant to ask was whether the #283 is obsolete after merging this PR?

**Comment by matthewmcgarvey** [comment]
O, I have definitely not covered all that is covered in that PR.

Things left to do:
- Nested destructuring of enums (in that PR)
- Nested destructuring of arrays (in that PR)
- Limited nested destructuring of tuples within wheres/statements (not in that PR)

**Comment by Sija** [comment]
@matthewmcgarvey Thanks, got it!
